### PR TITLE
allow trailing whitespace after flags

### PR DESCRIPTION
**Pull request by gbaz**
Resolves https://github.com/haskell/cabal/issues/7279

We just weren't lexing trailing whitespace in flag names. So a constraint like "pkg +flag , ..." would fail while "pkg +flag, ..." would succeed.

* [ ] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [ ] The documentation has been updated, if necessary.
* [x] Tests

### CODE REVIEW

**Review by Mikolaj — APPROVED** [review]
Wow, seriously.

**Comment by Mikolaj** [comment]
No idea why doctests in Quick jobs fail.

File: Cabal-syntax/src/Distribution/Types/Flag.hs [hunk 1]
```diff
@@ -111,7 +111,7 @@ instance Parsec FlagName where
     -- cabal check will do that.
     parsec = mkFlagName . lowercase <$> parsec'
       where
-        parsec' = (:) <$> lead <*> rest
+        parsec' = (<* P.skipSpaces) $ (:) <$> lead <*> rest
```
Comment (phadej) [inline Cabal-syntax/src/Distribution/Types/Flag.hs:114]: IMHO, wrong place. This is better fixed in `FlagAssignment` parser (`Parsec` instance).

(We don't skip spaces after package names, etc. etc)

Comment (gbaz) [reply Cabal-syntax/src/Distribution/Types/Flag.hs:114]: fair enough. this seems to trip some tests too. i'll take another look.

**Comment by phadej** [comment]
`doctest`s fails because when parsing `+foo -bar` the `spaces` added eats the space in between (after the `foo`), but the `FlagAssignment` parser looks for a space. (Because `+foo-bar` is something completely different!)

**Comment by phadej** [comment]
In fact, I wouldn't even touch the `FlagAssignment`, as [`parsecCommaList`](https://github.com/haskell/cabal/blob/master/Cabal-syntax/src/Distribution/Parsec.hs#L294-L295) already adds the `spaces` after the element.

One more reason to prioritize complete removal of `ReadP` stuff.

TL;DR this looks like a problem is in `cabal-install`, don't "fix" it in `Cabal`.

**Comment by gbaz** [comment]
Right. The FlagAssignment parser can't handle "+foo +bar " with a trailing space. It consumes the trailing space and since it treats it as a separator, immediately requires another flag. I've changed the approach taken there slightly, and I think it will work.

File: Cabal-syntax/src/Distribution/Types/Flag.hs [hunk 2]
```diff
@@ -290,16 +291,17 @@ parsecFlagAssignment = mkFlagAssignment <$>
...
 -- @since 3.4.0.0
 parsecFlagAssignmentNonEmpty :: CabalParsing m => m FlagAssignment
-parsecFlagAssignmentNonEmpty = mkFlagAssignment . toList <$>
-    P.sepByNonEmpty (onFlag <|> offFlag) P.skipSpaces1
+parsecFlagAssignmentNonEmpty = mkFlagAssignment <$> some (onFlag <|> offFlag)
   where
     onFlag = do
         _ <- P.char '+'
         f <- parsec
+        _ <- P.skipSpaces
```
Comment (phadej) [inline Cabal-syntax/src/Distribution/Types/Flag.hs:299]: this allows to parse `+foo+bar`, without a space, doesn't it?

Comment (gbaz) [reply Cabal-syntax/src/Distribution/Types/Flag.hs:299]: I guess it does parse that as the foo flag followed by the bar flag. I personally think that's fine, if a bit quirky. If you have a different suggestion as to how to fix the original bug, which prevents valid parses, I'm open to other ideas.

Comment (phadej) [reply Cabal-syntax/src/Distribution/Types/Flag.hs:299]:
```diff
diff --git a/Cabal-syntax/src/Distribution/Types/Flag.hs b/Cabal-syntax/src/Distribution/Types/Flag.hs
index 713430ee6..8346925aa 100644
--- a/Cabal-syntax/src/Distribution/Types/Flag.hs
+++ b/Cabal-syntax/src/Distribution/Types/Flag.hs
@@ -261,6 +261,14 @@ instance Pretty FlagAssignment where
 -- >>> simpleParsec "+foo -bar baz" :: Maybe FlagAssignment
 -- Nothing
 --
+-- Issue #7279 was fixed in Cabal-3.8
+--
+-- >>> explicitEitherParsec (parsecCommaList parsec) "+foo , -bar" :: Either String [FlagAssignment]
+-- Right [fromList [(FlagName "foo",(1,True))],fromList [(FlagName "bar",(1,False))]]
+--
+-- >>> simpleParsec "+foo+foo" :: Maybe FlagAssignment
+-- Nothing
+--
 -- @since 3.4.0.0
 --
 instance Parsec FlagAssignment where
@@ -272,19 +280,23 @@ dispFlagAssignment = Disp.hsep . map (Disp.text . showFlagValue) . unFlagAssignm
 
 -- | Parses a flag assignment.
 parsecFlagAssignment :: CabalParsing m => m FlagAssignment
-parsecFlagAssignment = mkFlagAssignment <$> many (onFlag <|> offFlag)
+parsecFlagAssignment = mkFlagAssignment <$> sepByEnding (onFlag <|> offFlag) P.skipSpaces1
   where
     onFlag = do
         _ <- P.char '+'
         f <- parsec
-        _ <- P.skipSpaces
         return (f, True)
     offFlag = do
         _ <- P.char '-'
         f <- parsec
-        _ <- P.skipSpaces
         return (f, False)
 
+    sepByEnding :: CabalParsing m => m a -> m b -> m [a]
+    sepByEnding p sep = afterSeparator where
+        element        = (:) <$> p <*> afterElement
+        afterElement   = sep *> afterSeparator <|> pure []
+        afterSeparator = element <|> pure []
+
 -- | Parse a non-empty flag assignment
 --
 -- The flags have to explicitly start with minus or plus.
```

File: Cabal-syntax/src/Distribution/Compat/CharParsing.hs [hunk 3]
```diff
@@ -355,3 +356,7 @@ munch = many . satisfy
 skipSpaces1 :: CharParsing m => m ()
 skipSpaces1 = skipSome space
 {-# INLINE skipSpaces1 #-}
+
+skipSpaces :: CharParsing m => m ()
```
Comment (phadej) [inline Cabal-syntax/src/Distribution/Compat/CharParsing.hs:360]: Remove. There is `spaces :: CharParsing m => m ()`.

File: Cabal-syntax/src/Distribution/Types/Flag.hs (hunk 2 above)
Comment (gbaz) [reply Cabal-syntax/src/Distribution/Types/Flag.hs]: Nice 👍

File: Cabal-syntax/src/Distribution/Compat/CharParsing.hs (hunk 3 above)
Comment (Mikolaj) [reply Cabal-syntax/src/Distribution/Compat/CharParsing.hs]: @gbaz: what about that one?

Comment (Mikolaj) [reply Cabal-syntax/src/Distribution/Compat/CharParsing.hs]: oh, I'm sorry, silly github has shown me outdated patch

**Review by Mikolaj — APPROVED** [review]
As I said, LGTM.

**Comment by jneira** [comment]
@MergifyIO rebase master

**Comment by jneira** [comment]
err, it needed a changelog i guess?

**Comment by Mikolaj** [comment]
> err, it needed a changelog i guess? facepalm

Right, we missed it.

**Comment by Mikolaj** [comment]
And a test.

File: Cabal-syntax/src/Distribution/Types/Flag.hs [hunk 4]
```diff
@@ -261,6 +261,17 @@ instance Pretty FlagAssignment where
...
 --
+-- Issue #7279 was fixed in Cabal-3.8
+--
+-- >>> explicitEitherParsec (parsecCommaList parsec) "+foo , -bar" :: Either String [FlagAssignment]
+-- Right [fromList [(FlagName "foo",(1,True))],fromList [(FlagName "bar",(1,False))]]
+--
+-- >>> explicitEitherParsec (parsecCommaList parsecFlagAssignmentNonEmpty) "+foo , -bar" :: Either String [FlagAssignment]
+-- Right [fromList [(FlagName "foo",(1,True))],fromList [(FlagName "bar",(1,False))]]
+--
+-- >>> simpleParsec "+foo+foo" :: Maybe FlagAssignment
```
Comment (phadej) [inline Cabal-syntax/src/Distribution/Types/Flag.hs:272]: @Mikolaj I hope that `doctest`s are still run as part of CI. i.e. there are tests.

Comment (Mikolaj) [reply Cabal-syntax/src/Distribution/Types/Flag.hs:272]: Oh, good, So that confirms what I hypothesised in #8040.

Comment (jneira) [reply Cabal-syntax/src/Distribution/Types/Flag.hs:272]: it does https://github.com/haskell/cabal/runs/5499562852?check_suite_focus=true

afaiu tests will prove this change does not break anything else, but they dont ensure we can break again flags + trailing whitespace

Comment (jneira) [reply Cabal-syntax/src/Distribution/Types/Flag.hs:272]: oh sorry doctests are inlined and 

`>>> explicitEitherParsec (parsecCommaList parsecFlagAssignmentNonEmpty) "+foo , -bar" :: Either String [FlagAssignment]
-- Right [fromList [(FlagName "foo",(1,True))],fromList [(FlagName "bar",(1,False))]]`

is testing this
